### DISCUSSION TITLE
ngx-mat-select-search, bootstrap and angular material compatibility

### DISCUSSION THREAD

**Discussion by ahmaddev1111** [TDRA-ae/aegov-dls#44] [Q&A] [answered]
Hi everyone,

I’m currently using ngx-mat-select-search in my Angular application and would like to integrate the UAE Design System. However, I’m concerned about potential CSS conflicts between the two.

Could anyone advise on the best way to achieve this integration? Also, are the styles from the UAE Design System compatible with Bootstrap and Angular Material?

Thank you for your help!

**Comment by moiz-conceptlz** [comment]
Hello @ahmaddev1111 

Noticed that the "ngx-mat-select-search" package classes are all prefixed with **"mat-mdc-select-*"** - hence I doubt there is a CSS conflict. Infact, you may have to consider the fact that you may additionally have to write CSS to ensure this component blends with the look and feel of the design system. It seems this component was strictly designed to work with Material UI only.

<img width="1165" alt="Screenshot 2024-07-16 at 1 28 46 PM" src="https://github.com/user-attachments/assets/d844d53e-cfaf-4fed-aabe-7d2ea2c07171">

Also noticed on https://github.com/bithost-gmbh/ngx-mat-select-search - that there is a note that says "This project is meant as a temporary implementation of https://github.com/angular/components/issues/5697." - Are you sure you want to use this in a production environment?

Also, in regards to __"are the styles from the UAE Design System compatible with Bootstrap and Angular Material?"__ 

The UAE design system is a design library in itself, and I'm not sure what you mean by compatible with Bootstrap. Bootstrap is a CSS library with its own set of design style. The purpose of a design system is to replace things like Bootstrap, and build a consistent use CSS library that can be used as the primary source for development of UI. However, to help developers create reusable code, the UAE design system utilizes TailwindCSS as a utility library.

Because TailwindCSS is the underlying utility library, nothing stops anyone from utilizing it with Angular. You can follow https://tailwindcss.com/docs/guides/angular to learn more about how TailwindCSS and the UAE Design System plugin can be used with Angular too. However, Material Angular is again a design system with it's own set of style UI components. If you need the functionality of the components defined in Material Angular, and it a requirement to use it, you will have to work to override classes to match the design system. You would follow https://material.angular.io/guide/customizing-component-styles as a guide on customizing the look of components.

Hope this was helpful.

**Reply by ahmaddev1111** [reply]
Hello Moiz,

Thank you for responding. My question revolves around implementing TDRA Guidelines in an existing application that has evolved over years, utilizing UI component libraries such as Bootstrap and Angular Material UI. These libraries include numerous pre-built components that enhance user experience, such as a searchable select component for the nationality dropdown in forms. This approach ensures usability by allowing users to filter and select their nationality from a list of countries, which aligns well with the goals of the TDRA design system.

How can we achieve this using the TDRA design system?

Your assistance is greatly appreciated!

Best regards,

**Reply by moiz-conceptlz** [reply, marked as answer]
@ahmaddev1111 I understand what you are referring to, and it's valid - existing use of components and libraries cannot be instantly changed. There are dependencies and things are working as intended!

Firstly; let me clarify that the design system, in its current version - including the guidelines, blocks and components has been designed to provide FGE's the ability to create consistent public-facing websites.

The purpose is to have consistent-looking, compliant, optimised websites with a consistent way to communicate information and be a source of information for the end-user to inform them about the interaction and possibilities with the entity. Using the design system and following the practices listed, aims to create a singular set of experience for the end-user to interact with government.

FGE's may also have functional applications, service applications, backend dashboards, service portals, applications that allow users to interact with the entity and more. Essentially - everything that happens after a LOGIN. The design system does not restrict the use of any library, use of a JS framework or any other method to create and manage these applications. 

I also understand that these portals and applications depend on these components and frameworks and cannot simply change. The easiest turnaround to match the look of the design system is to override their CSS.

All components currently in the design system are eventually CSS driven!  

Take an example of a solid primary button. The output CSS is;
```
.aegov-btn {
    display: inline-flex;
    flex-shrink: 0;
    cursor: pointer;
    -webkit-user-select: none;
    -moz-user-select: none;
    user-select: none;
    flex-wrap: wrap;
    align-items: center;
    justify-content: center;
    text-align: center;
    text-decoration-line: none;
    transition-property: color,background-color,border-color,text-decoration-color,fill,stroke,opacity,box-shadow,transform,filter,-webkit-backdrop-filter;
    transition-property: color,background-color,border-color,text-decoration-color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;
    transition-property: color,background-color,border-color,text-decoration-color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter,-webkit-backdrop-filter;
    transition-duration: .2s;
    transition-timing-function: cubic-bezier(.4,0,.2,1);
    height: 3rem;
    gap: .5rem;
    border-radius: .5rem;
    border-width: 2px;
    border-color: transparent;
    padding-left: 1.5rem;
    padding-right: 1.5rem;
    font-size: 1rem;
    line-height: 1.5rem;
    font-weight: 500;
    --tw-bg-opacity: 1;
    background-color: rgb(146 114 42 / var(--tw-bg-opacity));
    --tw-text-opacity: 1;
    color: rgb(255 255 255 / var(--tw-text-opacity))
}

.aegov-btn:focus {
    outline: 2px solid transparent;
    outline-offset: 2px
}

.aegov-btn:focus-visible {
    --tw-ring-opacity: 1;
    --tw-ring-color: rgb(250 212 79 / var(--tw-ring-opacity))
    --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
    --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(2px + var(--tw-ring-offset-width)) var(--tw-ring-color);
    box-shadow: var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow, 0 0 #0000);
    --tw-ring-offset-width: 2px
}

.aegov-btn:hover {
    --tw-bg-opacity: 1;
    background-color: rgb(182 138 53 / var(--tw-bg-opacity));
    --tw-text-opacity: 1;
    color: rgb(249 247 237 / var(--tw-text-opacity));
    --tw-shadow: 0px 0px 0px 6px;
    --tw-shadow-colored: 0px 0px 0px 6px var(--tw-shadow-color);
    box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
    --tw-shadow-color: #F2ECCF;
    --tw-shadow: var(--tw-shadow-colored)
}

.aegov-btn:disabled {
    pointer-events: none;
    cursor: not-allowed;
    opacity: .3;
    --tw-shadow: 0 0 #0000 !important;
    --tw-shadow-colored: 0 0 #0000 !important;
    box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow)!important
}
```

Overriding the CSS for an existing component to match this can be a quick solution.
